Add party without relationship

laura-barluzzi opened the pull request:
### Proposed changes in this pull request

This PR addresses the issue _Ability to add parties without relationships_ #1581. This PR doesn't change existing functionalities, it just adds one. 

This PR is largely based on existing code that is currently unused in the platform. I have patched what I found together and I believe it works fine. 

1. I added the + Add button in the template `party/party_list.html`
<img src="https://user-images.githubusercontent.com/16849118/28437952-55a8cb8e-6d51-11e7-97ba-be07770943d6.png" width="300">

2. I added the + Add party option into the `single project header dropdown` as well, along with `add location` and `add resource`.
3. When clicked, the buttons send you to an existing (but still not used) url path that brings to `party/forms.PartiesAdd` view and template `party/party_add.html`
3. I changed template `party_add.html` to be like `party_edit.html` so that it displays a modal
<img src="https://user-images.githubusercontent.com/16849118/28437894-24dffcf2-6d51-11e7-8354-6be3434fd99d.png" width="300">

5. Currently, both `party_edit.html` and `party_add.html` include template `party/party_form.html` in the body of their modals
6. I modified a bit `party/party_form.html` template, since before was not displaying a dropdown menu for party types. Now it displays the dropdown:
<img src="https://user-images.githubusercontent.com/16849118/28438076-ba5baeca-6d51-11e7-9acf-8d5a9c162fba.png" width="300">

7. I added **`{% if is_allowed_add_party %}`** around the two buttons in order to display them only if the current user can add a party themselves.

### Summary
This PR allows users to add a new party (with `name` and `party type` fields) without associating the party to any relationships (n.b. users can still add a party through the "traditional" process as well). If the xlsx form is downloaded, it will add the party-without-relationship information _**only**_ in the `parties` label. 
Hereafter the examples of `parties` and `relationships` labels of a downloaded xlsx form when there is a party named "hhh" without a relationship.
<img src="https://user-images.githubusercontent.com/16849118/28444216-76802eaa-6d70-11e7-9719-587879963140.png" width="300"><img src="https://user-images.githubusercontent.com/16849118/28444297-068d1f08-6d71-11e7-9c0a-09ad0c14fab7.png" width="300">

### When should this PR be merged
- After it has been reviewed
- After it has been approved

### Risks
- Low

### Follow-up actions
- Double check if this meets the client's demand

### Checklist (for reviewing)

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.
  - [ ] Review 1
  - [ ] Review 2

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
  - [ ] Review 1
  - [ ] Review 2
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 
  - [ ] Review 1
  - [ ] Review 2

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
  - [ ] Review 1
  - [ ] Review 2
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
  - [ ] Review 1
  - [ ] Review 2
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
  - [ ] Review 1
  - [ ] Review 2
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
  - [ ] Review 1
  - [ ] Review 2
- **Has the scalability of this change been evaluated?**
  - [ ] Review 1
  - [ ] Review 2
- **Is there a maintenance plan in place?**
  - [ ] Review 1
  - [ ] Review 2

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
  - [ ] Review 1
  - [ ] Review 2
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
  - [ ] Review 1
  - [ ] Review 2
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?
  - [ ] Review 1
  - [ ] Review 2

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
  - [ ] Review 1
  - [ ] Review 2
- **Are all UI and API inputs run through forms or serializers?** 
  - [ ] Review 1
  - [ ] Review 2
- **Are all external inputs validated and sanitized appropriately?**
  - [ ] Review 1
  - [ ] Review 2
- **Does all branching logic have a default case?**
  - [ ] Review 1
  - [ ] Review 2
- **Does this solution handle outliers and edge cases gracefully?**
  - [ ] Review 1
  - [ ] Review 2
- **Are all external communications secured and restricted to SSL?**
  - [ ] Review 1
  - [ ] Review 2

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 
  - [ ] Review 1
  - [ ] Review 2

clash99 commented on cadasta/templates/party/party_list.html:
```diff
@@ -15,6 +15,11 @@
       </div>
       <div class="panel panel-default">
         <div class="panel-body">
+          <div class="top-btn pull-right">
```
Add top-add class to this line

clash99 commented on cadasta/templates/party/party_list.html:
```diff
@@ -15,6 +15,11 @@
       <div class="panel panel-default">
         <div class="panel-body">
+          <div class="top-btn pull-right">
+            <a href="{% url 'parties:add' object.organization.slug object.slug %}" class="btn btn-primary add-org">
```
Remove add-org class and add btn-sm class

clash99 commented on cadasta/templates/party/party_form.html:
```diff
@@ -3,25 +3,15 @@
 {% csrf_token %}
 <div class="form-group{% if form.name.errors %} has-error{% endif %}">
-  <label class="control-label" for="{{ form.name.id_for_label }}" {{ form.name.field.labels_xlang|safe }}>{% trans "Name" %}</label>
+  <label class="control-label required" for="{{ form.name.id_for_label }}" {{ form.name.field.labels_xlang|safe }}>{% trans "Name" %}</label>
```
Field label should be "Party name"

clash99 commented on cadasta/templates/party/party_form.html:
```diff
@@ -3,25 +3,15 @@
 </div>
 
-{% if form.new_item %}
 <div class="form-group{% if form.type.errors %} has-error{% endif %}">
```
Field label should be "Party type" and it should be required

clash99 commented on cadasta/templates/party/party_add.html:
```diff
@@ -26,28 +28,33 @@
             {% include "party/party_form.html" %}
-
+          </div>
+          <div class="modal-footer">
```
Can we return to party index, not the party detail, after adding a new party here?

clash99 replied:
When adding a resource to a project from the resource index page or from the project header, you are given a modal and then once complete, it drops user back to resource index. I was thinking to mirror that flow.

laura-barluzzi commented:
@clash99 here some screenshots of the new version of the Add Party modal. They show (1) the first view of the modal once opened; (2) View of error messages of _both_ fields and (3) the current dropdown design that I copied-and-pasted by the existing one in use in the `relationship_add.html` modal.

<img src="https://user-images.githubusercontent.com/16849118/28479486-80b5e986-6e11-11e7-8828-78608619f352.png" height="150">     <img src="https://user-images.githubusercontent.com/16849118/28479488-80b8db82-6e11-11e7-9f6b-8d6a402e9a52.png" height="150">
<img src="https://user-images.githubusercontent.com/16849118/28479487-80b79236-6e11-11e7-878f-16012cb3b260.png" height="150">

laura-barluzzi commented:
Thank @oliverroick for your feedback.

1. I added the option to add a party into the header dropdown. Now a user can add a party in both ways -- from the dropdown, or from the Party index page. I think it's totally fine to keep them both.
2. About where to redirect the user **_after adding a new party_** (`Party Detail` vs `Party List`) I want to wait for @clash99 opinion. I don't have a strong opinion about this and I think that either way is fine. I wait for a final decision to be taken before changing the code.
3. I added the condition wrapper to check **if the current user is allowed to add the party** in order to display the respective buttons. I just added the permission `party.create` in a pre-existing mixins class that handles permissions.

Hereafter there are some screenshots of the `Add party` buttons displayed based on user's roles. On the left is the header dropdown, on the right the button in the Party Index.
_clemens's role = **project user**_
<img src="https://user-images.githubusercontent.com/16849118/28549758-49924ae4-7091-11e7-9721-57efa655f11d.png" height="100"> - <img src="https://user-images.githubusercontent.com/16849118/28549786-7a718788-7091-11e7-8a50-fb40f6a0f1cd.png"
 height="100">

_clemens's role = **data collector**_
<img src="https://user-images.githubusercontent.com/16849118/28549435-dc6f84ce-708e-11e7-940c-4a5ad0fc004f.png" height="100"> - <img src="https://user-images.githubusercontent.com/16849118/28549935-68dc1064-7092-11e7-94eb-e711dbba10a5.png" height="100">

_clemens's role = **project manager**_
<img src="https://user-images.githubusercontent.com/16849118/28550005-eb2b453a-7092-11e7-9365-9c85c6d62103.png" height="100"> - <img src="https://user-images.githubusercontent.com/16849118/28549975-b9d06f4c-7092-11e7-8a8a-bc9d2d9ccf1a.png" height="100">

_laura's **organization admin**_
<img src="https://user-images.githubusercontent.com/16849118/28549521-7904ea86-708f-11e7-9772-85e1f954627f.png" height="100"> - <img src="https://user-images.githubusercontent.com/16849118/28549860-e3168d60-7091-11e7-8050-715ac7dbb23a.png" height="100">

clash99 commented:
@laura-barluzzi - Its fine to drop-off at the party detail page. My reasoning for returning to the index was that the party detail page is a dead-end right now until some small changes are made but those are in another task so I can deal with that then. For simplicity's sake, just keep it as-is.

oliverroick commented:
@laura-barluzzi I discussed the redirecting problem with Chandra earlier today and we decided to redirect to the list page. 

I'll review your changes tomorrow morning, it should be good to go soon.

clash99 commented on cadasta/templates/organization/project_wrapper.html:
```diff
@@ -76,6 +76,7 @@ <h1 class="{% if is_administrator or is_allowed_add_location or is_allowed_add_r
           </button>
           <ul class="dropdown-menu">
             {% if is_allowed_add_resource %}<li><a href="{% url 'resources:project_add_existing' object.organization.slug object.slug %}">{% trans "Add resource" %}</a></li>{% endif %}
+            {% if is_allowed_add_party %}<li><a href="{% url 'parties:add' object.organization.slug object.slug %}">{% trans "Add party" %}</a></li>{% endif %}
```
Sorry - one small change. Can you put "Add party" before "Add resource" to mirror the left subnav?

laura-barluzzi commented:
@clash99 I switched the location of add party and add resource in the dropdown. I amended the last commit so you see it inside the pass tests commit.

laura-barluzzi commented:
@oliverroick @clash99 ready to be merged.
